source/agora/crypto/Hash.d

```diff
@@ -120,60 +119,32 @@ public Hash hashFull (T) (scope const auto ref T record)
 /// Ditto
 public void hashPart (T) (scope const auto ref T record, scope HashDg state)
     /*pure*/ nothrow @nogc
-    if (is(T == struct))
 {
-    static if (is(typeof(T.init.computeHash(HashDg.init))))
-        record.computeHash(state);
-    else static if (__traits(compiles, () { const ubyte[] r = T.init[]; }))
-        state(record[]);
+    static if (is(T == struct))
+    {
+        static if (is(typeof(T.init.computeHash(HashDg.init))))
+            record.computeHash(state);
+        else
+            foreach (const ref field; record.tupleof)
+                hashPart(field, state);
+    }
     else
-        foreach (const ref field; record.tupleof)
-            hashPart(field, state);
-}
-
-/// Ditto
-public void hashPart (ubyte record, scope HashDg state) /*pure*/ nothrow @nogc @trusted
-{
-    state((cast(ubyte*)&record)[0 .. ubyte.sizeof]);
-}
-
-/// Ditto
-public void hashPart (ushort record, scope HashDg state) /*pure*/ nothrow @nogc @trusted
-{
-    state(nativeToLittleEndian(record)[0 .. ushort.sizeof]);
-}
-
-/// Ditto
-public void hashPart (uint record, scope HashDg state) /*pure*/ nothrow @nogc @trusted
-{
-    state(nativeToLittleEndian(record)[0 .. uint.sizeof]);
-}
-
-/// Ditto
-public void hashPart (ulong record, scope HashDg state) /*pure*/ nothrow @nogc @trusted
-{
-    state(nativeToLittleEndian(record)[0 .. ulong.sizeof]);
-}
-
-/// Ditto
-public void hashPart (in char[] record, scope HashDg state) /*pure*/ nothrow @nogc @trusted
-{
-    state(cast(const ubyte[])record);
-}
-
-/// Ditto
-public void hashPart (in ubyte[] record, scope HashDg state)
-    /*pure*/ nothrow @nogc @safe
-{
-    state(record);
+        doHashWithSerializer(record, state);
 }
 
-/// Ditto
-public void hashPart (T) (in T[] records, scope HashDg state)
-    /*pure*/ nothrow @nogc @safe
+// We hash after binary serialization to prevent hash collisions related to empty arrays
+// See issue https://github.com/bpfkorea/agora/issues/1331
+private static void doHashWithSerializer (T) (scope const auto ref T record, scope HashDg dg)
+    nothrow @nogc @trusted
 {
-    foreach (ref record; records)
-        hashPart(record, state);
+    // Enable the serializePart to be called even though it is not `@nogc nothrow`
+    alias Expected = void delegate() @safe nothrow @nogc;
+    void fakeAttributes () @safe
+    {
+        serializePart(record, dg, CompactMode.No);
```

Comment:
This is also no bueno. We want to use `CompactMode.No`, **but** it means the hashes are platform dependent (you'll get a different hash on 32 bits). So that might not work out the way we thought.

Reply:
I thought we were using `CompactMode.No` for that very reason, to make it platform independent.

Reply:
Updating to use `CompactMode.Yes`